### PR TITLE
Fix transaction validation.

### PR DESCRIPTION
**Pull request by arijitAD**
## Changes

- Fix balance key in storage.
- Add balance transfer test.

## Tests

```
TestAuthorSubmitExtrinsic
TestNodeRuntime_ValidateTransaction
```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [x] All integration tests and required coverage checks are passing

## Issues

- #1172

### CODE REVIEW

File: lib/runtime/wasmer/exports_test.go [hunk 1]
```diff
@@ -119,6 +120,44 @@ func TestInstance_Version_NodeRuntime(t *testing.T) {
...
+	return append(append(append(h0, h1...), h2...), pub...)
+}
+
+func TestNodeRuntime_ValidateTransaction(t *testing.T) {
+	DefaultTestLogLvl = 4
+
+	alicePub := common.MustHexToBytes("0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d")
+	aliceBalanceKey := balanceKey(t, alicePub)
+
+	rt := NewTestInstance(t, runtime.NODE_RUNTIME)
```
Comment (noot) [inline lib/runtime/wasmer/exports_test.go:139]: you can change this to load the gssmr genesis:
```
	gen, err := genesis.NewGenesisFromJSONRaw("../../../chain/gssmr/genesis-raw.json")
	require.NoError(t, err)

	genTrie, err := genesis.NewTrieFromGenesis(gen)
	require.NoError(t, err)

	// set state to genesis state
	genState := storage.NewTestTrieState(t, genTrie)

	cfg := &Config{}
	cfg.Storage = genState
	cfg.LogLvl = 5

	instance, err := NewRuntimeFromGenesis(gen, cfg)
	require.NoError(t, err)
```

File: tests/rpc/rpc_02-author_test.go [hunk 2]
```diff
@@ -17,15 +17,133 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	// Create the extrinsic
+	ext := types.NewExtrinsic(c)
+
+	// TODO: the genesis hash is wrong.
+	// genesisHash, err := api.RPC.Chain.GetBlockHash(0)
+	// require.NoError(t, err)
```
Comment (noot) [inline tests/rpc/rpc_02-author_test.go:77]: then you can uncomment this

File: tests/rpc/rpc_02-author_test.go [hunk 3]
```diff
@@ -17,15 +17,133 @@
...
+	key, err := types.CreateStorageKey(meta, "System", "Account", signature.TestKeyringPairAlice.PublicKey, nil)
+	require.NoError(t, err)
+
+	var nonce uint32
+	ok, err := api.RPC.State.GetStorageLatest(key, &nonce)
+	require.NoError(t, err)
+	require.True(t, ok)
+
+	// this is the actual genesis hash for GenesisDefault
+	genesisHash := types.NewHash(common.MustHexToBytes("0x03170a2e7597b7b7e3d84c05391d139a62b157e78786d8c082f29dcf4c111314"))
```
Comment (noot) [inline tests/rpc/rpc_02-author_test.go:91]: and remove this, then use the extrinsic from this test in the unit test, and it should work

File: lib/runtime/wasmer/exports_test.go (hunk 1 above)
Comment (arijitAD) [reply lib/runtime/wasmer/exports_test.go:139]: The test is running without the genesis. The balance data was incorrect. I updated it to `0x000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000`

File: chain/gssmr/config.toml [hunk 4]
```diff
@@ -33,8 +33,11 @@ nobootstrap = false
 nomdns = false
 
 [rpc]
-enabled = false
+enabled = true
+external = false
```
Comment (noot) [inline chain/gssmr/config.toml:37]: can you change this file back? we don't want rpc/ws on by default

File: dot/rpc/modules/author.go [hunk 5]
```diff
@@ -40,7 +66,9 @@ type AuthorModule struct {
 type KeyInsertRequest []string
 
 // Extrinsic represents a hex-encoded extrinsic
-type Extrinsic string
+type Extrinsic struct {
+	Data string
+}
```
Comment (noot) [inline dot/rpc/modules/author.go:71]: this changes the extrinsic format in JSON, is this related to the polkadot-js issues?

File: lib/runtime/wasmer/exports_test.go [hunk 6]
```diff
@@ -119,6 +121,56 @@ func TestInstance_Version_NodeRuntime(t *testing.T) {
 	require.Equal(t, expected, version.RuntimeVersion)
 }
 
+func balanceKey(t *testing.T, pub []byte) []byte {
```
Comment (noot) [inline lib/runtime/wasmer/exports_test.go:124]: this function should be used by the build-spec command to turn the human-readable genesis into a raw genesis, but that could be done separately

File: dot/types/account.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package types
+
+import gtypes "github.com/centrifuge/go-substrate-rpc-client/v2/types"
+
+// AccountInfo Information of an account.
+type AccountInfo struct {
+	// The number of transactions this account has sent.
+	Nonce gtypes.U32
```
Comment (noot) [inline dot/types/account.go:8]: it might be nicer to have this as uint32 so we don't need to import this package every time we want to use this type

File: dot/types/account.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+	// The number of other modules that currently depend on this account's existence. The account
+	// cannot be reaped until this is zero.
+	RefCount gtypes.U32
+	// The additional data that belongs to this account. Used to store the balance(s) in a lot of chains.
+	Data struct {
+		Free       gtypes.U128
```
Comment (noot) [inline dot/types/account.go:14]: same here, we can change this to our uint128 type (you will probably need to add `Encode` and `Decode` for it which is the number in 16-byte little endian format)

File: tests/rpc/rpc_02-author_test.go [hunk 9]
```diff
@@ -22,10 +22,88 @@ import (
...
+	gsrpc "github.com/centrifuge/go-substrate-rpc-client/v2"
+	"github.com/centrifuge/go-substrate-rpc-client/v2/signature"
+	"github.com/centrifuge/go-substrate-rpc-client/v2/types"
 	"github.com/stretchr/testify/require"
 )
 
+func TestAuthorSubmitExtrinsic(t *testing.T) {
+	if utils.MODE != rpcSuite {
+		_, _ = fmt.Fprintln(os.Stdout, "Going to skip RPC suite tests")
+		//return
```
Comment (noot) [inline tests/rpc/rpc_02-author_test.go:36]: uncomment?

File: tests/rpc/rpc_02-author_test.go [hunk 10]
```diff
@@ -22,10 +22,88 @@ import (
...
+	require.True(t, ok)
+
+	o := types.SignatureOptions{
+		BlockHash:          genesisHash,
+		Era:                types.ExtrinsicEra{IsImmortalEra: true},
+		GenesisHash:        genesisHash,
+		Nonce:              types.NewUCompactFromUInt(uint64(nonce)),
+		SpecVersion:        rv.SpecVersion,
+		Tip:                types.NewUCompactFromUInt(0),
+		TransactionVersion: 1, // TODO: rv.TransactionVersion == 0 but runtime expects 1
```
Comment (noot) [inline tests/rpc/rpc_02-author_test.go:93]: can you open an issue for this?

**Review by edwardmack — APPROVED** [review]
Commented on wrong pr.

File: dot/rpc/modules/author.go (hunk 5 above)
Comment (arijitAD) [reply dot/rpc/modules/author.go:71]: This is needed for unmarshalling the JSON request. This was broken but it didn't error out as the test was skipped.

File: tests/rpc/rpc_02-author_test.go (hunk 10 above)
Comment (arijitAD) [reply tests/rpc/rpc_02-author_test.go:91]: Created an issue for this. #1337

File: lib/runtime/wasmer/exports_test.go (hunk 6 above)
Comment (arijitAD) [reply lib/runtime/wasmer/exports_test.go:124]: Created an issue for this. #1339

**Review by RyRy79261 — COMMENTED** [review]
Not sure how to run the tests so ran `go test ./... -short` 3 times and the test failed at this: 
image

**Comment by arijitAD** [comment]
> Not sure how to run the tests so ran `go test ./... -short` 3 times and the test failed at this:
> image

I have fixed all the test. Since the test was running in docker, it was not able to find the subkey binary.

**Review by noot — APPROVED** [review]
great stuff!
